### PR TITLE
Use pip-compile-multi

### PR DESCRIPTION
**Pull request by rail**
I was playing with pip-compile-multi and signingscript's outdated compiled requirements-prod.txt file and ended up with something like this. Not quite sure if this is what we want, but I'll try. :)

I also added a check in order to verify that the compiled files are on par with their `.in` equivalents.

@escapewindow what do you think?

### CODE REVIEW

File: requirements/test.in [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+flake8
+flake8_docstrings
+mar
+mock
+pip-compile-multi
+pytest
+pytest-asyncio>=0.6.0
+pytest-cov
+pytest-mock
+scriptworker
```
Comment (escapewindow) [inline requirements/test.in]: we already have scriptworker in base.in, so this is a dup. Harmless though.

File: tox.ini [hunk 2]
```diff
@@ -14,26 +14,17 @@ passenv =
...
-    pytest-asyncio>=0.6.0
-    pytest-cov
-    pytest-mock
+    -r requirements/test.txt
 
 commands=
     py.test --cov-config .coveragerc --cov=signingscript --cov-report term-missing
     coverage html
     flake8
+    pip-compile-multi verify
```
Comment (escapewindow) [inline tox.ini]: We may not want to do this unless we restrict ourselves to a single version of python and a single host arch. (The generated pins and hashes can differ between python 3.6 and 3.7, or between linux and macosx). We should probably have a separate target for this test if we want to run it.

Instead, we should probably test against the unpinned deps, and have a separate process or test for pinning.

File: tox.ini [hunk 3]
```diff
@@ -14,26 +14,17 @@ passenv =
...
-    flake8
-    flake8_docstrings
-    mock
-    mar
-    scriptworker
-    pytest
-    pytest-asyncio>=0.6.0
-    pytest-cov
-    pytest-mock
+    -r requirements/test.txt
```
Comment (escapewindow) [inline tox.ini]: See the below comment.

File: tox.ini [hunk 4]
```diff
@@ -14,26 +14,17 @@ passenv =
...
     py.test --cov-config .coveragerc --cov=signingscript --cov-report term-missing
     coverage html
     flake8
+    pip-compile-multi verify
 
 [testenv:coveralls]
 deps=
-    python-coveralls
-    coverage>=4.2b1
+    -r requirements/coveralls.txt
```
Comment (escapewindow) [inline tox.ini]: See the above comment.

**Comment by escapewindow** [comment]
Hm, pulling from your branch, I was able to run tox for both py36 and py37 without issue on osx. I may be mistaken about my review comments above, though it's also possible we just don't have any dependencies currently that hit it, but we may in the future (or in other *scripts)

**Comment by rail** [comment]
> Hm, pulling from your branch, I was able to run tox for both py36 and py37 without issue on osx. I may be mistaken about my review comments above, though it's also possible we just don't have any dependencies currently that hit it, but we may in the future (or in other *scripts)

Actually I hit an issue, when I tried to run tox on 3.6 with requirements generated with 3.7. A couple of packages (second level deps iirc) were marked as for python<3.7, so they were not in the generated txt files. Regenerating the deps on 3.6 solved the issue. I'm not sure what would be the best approach here, tbh.

File: tox.ini (hunk 2 above)
Comment (rail) [reply tox.ini]: Hmm, I thought that the list of hashes contains all possible combinations (tarball, version/platform specific wheels) for that particular version...

Comment (rail) [reply tox.ini]: So I think if we run this on all supported python versions (just as it is now in the patch), we should be safe. I got red tox runs until I fixed the issue with my initial compilation that I made on py37, but it was not sufficient for py36.

Comment (escapewindow) [reply tox.ini]: I may be misunderstanding, but aiui:

- we land pinned hashes and versions
- these hashes and versions will be tied to a python minor version and OS arch
- if we run this test on a different python minor version and OS arch, we may hit issues

**Comment by escapewindow** [comment]
> > Hm, pulling from your branch, I was able to run tox for both py36 and py37 without issue on osx. I may be mistaken about my review comments above, though it's also possible we just don't have any dependencies currently that hit it, but we may in the future (or in other *scripts)
> 
> Actually I hit an issue, when I tried to run tox on 3.6 with requirements generated with 3.7. A couple of packages (second level deps iirc) were marked as for python<3.7, so they were not in the generated txt files. Regenerating the deps on 3.6 solved the issue. I'm not sure what would be the best approach here, tbh.

I believe that's because dependencies can be specified for a specific minor version of python, so 3.6 and 3.7 get different lists. And wheels can be compiled for different OS arches, so linux and macosx get different lists. My current thinking about this has been to rely on unpinned deps for testing, and running pinning if and when we have a specific OS/python minor version we're targeting. That would be for deployment or automation, rather than development.

**Comment by rail** [comment]
Let me remove that line then.

**Pull request merged**